### PR TITLE
gitserver: Add BatchLog

### PR DESCRIPTION
**Pull request by efritz**
**Reviewers**: Please review by commit. I've separated out the stuff that's just observability framework and separated code, test, and generated code additions to help aid in understanding of the new code.

This PR adds a `BatchLog` method to the gitserver client, which makes a minimal number of network requests to issue `git log <format> <commit>` commands for a set of repositories. This is a partial fix for #32328, which concerns the performance of certain cross-repository code intelligence queries. See the following partial trace for the dominating areas of the query:

**Bad, sequential network requests in trace**: Bad, sequential network requests in trace

This PR also provides full observability including metrics and tracing. A Honeycomb dataset would also be trivial to add on the server side as we used `internal/observation`, which supports all of our supported observability sinks today.

**Frontend dashboard**: Frontend dashboard

**Git Server dashboard**: Git Server dashboard

**(New) Local Jaeger trace**: Local Jaeger trace

## Test plan

Additional unit tests should provide good coverage (for confidence regarding batch construction and multiple request paths), and existing Code Intel QA tests exercise this path with exactly one gitserver instance (as a sanity check).

See a [main-dry-run](https://buildkite.com/sourcegraph/sourcegraph/builds/141117).

### CODE REVIEW

**Comment by tsenart** [comment]
FYI, we'd benefit from this in a very core part of search, repo resolution. Particularly when not search the default branch HEAD, we need to validate each revision being searched over. We do this in dependencies search for instance, so this could have a very good impact there too.

https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/search/repos/repos.go?L337

Another idea to speed this up is https://github.com/sourcegraph/sourcegraph/issues/28476

**Comment by ryanslade** [comment]
Nice, thanks for proposing this. 

Approach looks good and happy that we're targetting a specific use case rather than trying to make too general up front.

File: internal/gitserver/observability.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+		return observationContext.Operation(observation.Op{
+			Name: fmt.Sprintf("gitserver.client.%s", name),
+		})
+	}
+
+	return &operations{
+		batchLog:       op("BatchLog"),
+		batchLogSingle: subOp("batchLogSingle"),
+	}
+}
```
Comment (varungandhi-src) [inline internal/gitserver/observability.go:50]: Is it intentional that a bunch of this code is almost identical to the corresponding observability code in `gitserver/server/observability.go` itself, except for the name strings? It seems a little strange; will any future changes likely need to happen both in this file and in `gitserver/server/observability.go`?

File: cmd/gitserver/server/server.go [hunk 2]
```diff
@@ -1140,6 +1142,96 @@ func matchCount(cm *protocol.CommitMatch) int {
...
+func (s *Server) handleBatchLog(w http.ResponseWriter, r *http.Request) {
+	operations := s.ensureOperations()
+
+	// Run git log for a single repository.
+	// Invoked multiple times from the handler defined below.
+	performGitLogCommand := func(ctx context.Context, repoCommit api.RepoCommit, format string) (output string, isRepoCloned bool, err error) {
+		ctx, endObservation := operations.batchLogSingle.With(ctx, &err, observation.Args{
+			LogFields: []log.Field{
+				log.String("repoCommit.Repo", string(repoCommit.Repo)),
+				log.String("repoCommit.CommitID", string(repoCommit.CommitID)),
```
Comment (varungandhi-src) [inline cmd/gitserver/server/server.go:1154]: Should this be a method call on `repoCommit` that sits next to the type definition? I [ran into a similar problem earlier](https://github.com/sourcegraph/sourcegraph/pull/33122), where some fields were not logged because a type evolved over time but logging sites were not updated.

Also, I think there is a concern about us logging the repo name here without evidence that the repo name is public/OK to log.

File: cmd/gitserver/server/server.go [hunk 3]
```diff
@@ -1140,6 +1142,96 @@ func matchCount(cm *protocol.CommitMatch) int {
...
+
+		// Perform requests in each repository in the input batch. We do this synchronously
+		// today so that this endpoint remains simple. If it is determined that performing
+		// the git log requests in parallel _on each shard_ is necesesary, we should keep
+		// the following sequence of actions as simple as possible.
+
+		results := make([]protocol.BatchLogResult, 0, len(req.RepoCommits))
+		for _, repoCommit := range req.RepoCommits {
+			output, ok, err := performGitLogCommand(ctx, repoCommit, req.Format)
+			if err == nil && !ok {
```
Comment (varungandhi-src) [inline cmd/gitserver/server/server.go:1206]:
```suggestion
			output, isRepoCloned, err := performGitLogCommand(ctx, repoCommit, req.Format)
			if err == nil && !isRepoCloned {
```
nit: Matching the names on the results in the `performGitLogCommand` with the bindings here makes it easier to independently understanding the body of `performGitLogCommand` and the usage of the `ok` variable here.

File: cmd/gitserver/server/server.go [hunk 4]
```diff
@@ -1140,6 +1142,96 @@ func matchCount(cm *protocol.CommitMatch) int {
...
+			}})
+		}()
+
+		// Read request body
+		var req protocol.BatchLogRequest
+		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+			return http.StatusBadRequest, err
+		}
+		logger.Log(log.Int("req.numRepoCommits", len(req.RepoCommits)))
+		logger.Log(log.String("req.format", req.Format))
```
Comment (varungandhi-src) [inline cmd/gitserver/server/server.go:1196]: Same Q about moving the logged fields near the type definition as in https://github.com/sourcegraph/sourcegraph/pull/33195/commits/d136dd230a3aaca3811c0ba1d21e0bb97f8ed833#r839148869.

File: internal/gitserver/protocol/gitserver.go [hunk 5]
```diff
@@ -98,6 +98,28 @@ type ExecRequest struct {
...
 }
 
+// BatchLogRequest is a request to execute a `git log` command inside a set of
+// git repositories present on the target shard.
+type BatchLogRequest struct {
+	RepoCommits []api.RepoCommit `json:"repoCommits"`
+
+	// Format is the entire `--format=<format>` argument to git log. This value
+	// is expected to be non-empty.
+	Format string `json:"format"`
```
Comment (varungandhi-src) [inline internal/gitserver/protocol/gitserver.go:108]: Q: Is the validation for these fields deliberately omitted? IIUC the expectation is that  `len(req.RepoCommits) > 0 && strings.HasPrefix(req.Format, "--format=")` but these aren't being checked.

File: internal/gitserver/observability.go (hunk 1 above)
Comment (efritz) [reply internal/gitserver/observability.go:50]: They _happen_ to have the same operations today, but I would expect that the client and endpoint set will not necessarily be one-to-one in the future.

Also, we may want to explicit rename `batchLogSingle` to something less generic (I named the observability fields prior to refactors, so they're likely *bad* as they currently are), like `gitLogOneRepo` and `requestBatchFromShard` for the server and client, respectively.

File: cmd/gitserver/server/server.go (hunk 2 above)
Comment (efritz) [reply cmd/gitserver/server/server.go:1154]: > Should this be a method call on repoCommit that sits next to the type definition?

I like the pattern that you linked, but I think that applies more specifically to _options_ rather than these API types (which is named after the fields that compose it). I don't think there's risk of expanding field _in this particular case_, but if the uniformity is desirable I'm happy to add a method.

> I think there is a concern about us logging the repo name here without evidence that the repo name is public/OK to log.

Yeah I wanted to check around wayyy before pre-merge about these kind of things for new code in a core service. I followed the lead of what was already in that file, so I don't think I'm breaking _new_ rules, but might be reinforcing already broken ones.

cc @andreeleuterio about the PII concern in logs here

File: internal/gitserver/protocol/gitserver.go (hunk 5 above)
Comment (efritz) [reply internal/gitserver/protocol/gitserver.go:108]: > deliberately omitted

No, due to laze.

File: internal/gitserver/client.go [hunk 6]
```diff
@@ -608,6 +631,163 @@ var deadlineExceededCounter = promauto.NewCounter(prometheus.CounterOpts{
...
+		})
+		defer func() {
+			endObservation(1, observation.Args{
+				LogFields: []log.Field{
+					log.Int("numProcessed", numProcessed),
+				},
+			})
+		}()
+
+		uri := "http://" + addr + "/batch-log"
```
Comment (andreeleuterio) [inline internal/gitserver/client.go:666]: is `addr` user-controlled?

Comment (efritz) [reply internal/gitserver/client.go:666]: No, comes from `c.AddrForRepo` on L728.

File: cmd/gitserver/server/server.go (hunk 2 above)
Comment (varungandhi-src) [reply cmd/gitserver/server/server.go:1154]: > I don't think there's risk of expanding field in this particular case, but if the uniformity is desirable I'm happy to add a method.

I personally think that it would be valuable to follow it as a pattern in more parts of the codebase. I have not yet gone about gathering consensus on that. If we do gain broad consensus, we'll have to migrate a bunch of code to that pattern, and this would be a small bit of work in that large chunk of work.

I think it's OK to leave this as it is, but slight preference to change it because of the above. Your call.

> but might be reinforcing already broken ones.

Yep, that was my concern.

File: enterprise/internal/codeintel/gitserver/client.go [hunk 7]
```diff
@@ -89,6 +89,14 @@ func (c *Client) CommitsExist(ctx context.Context, commits []RepositoryCommit) (
...
 		exists[i] = e
 	}
+	if len(exists) != len(commits) {
+		// Add assertion here so that the blast radius of new or newly discovered errors southbound
+		// from the internal/vcs/git package does not leak into code intelligence. The existing callers
+		// of this method panic when this assertion is not met. Describing the error in more detail here
+		// will not cause destruction outside of the particular user-request in which this assertion
+		// was not true.
+		return nil, errors.Newf("expected slice returned from git.CommitsExist to have len %d, but has len %d", len(commits), len(exists))
+	}
```
Comment (varungandhi-src) [inline enterprise/internal/codeintel/gitserver/client.go:97]: IIUC, this check will never fire because you're doing `exists := make([]bool, len(commits))` instead of `exists := make([]bool, 0, len(commits))`, so `exists` will already have been zero-initialized and have length `len(commits)`.

File: internal/gitserver/client.go [hunk 8]
```diff
@@ -614,6 +631,163 @@ var deadlineExceededCounter = promauto.NewCounter(prometheus.CounterOpts{
...
+func repoNamesFromRepoCommits(repoCommits []api.RepoCommit) []string {
+	repoNameMap := make(map[api.RepoName]struct{}, len(repoCommits))
+	for _, rc := range repoCommits {
+		repoNameMap[rc.Repo] = struct{}{}
+	}
+	repoNames := make([]string, 0, len(repoNameMap))
+	for repoName := range repoNameMap {
+		repoNames = append(repoNames, string(repoName))
+	}
+	sort.Strings(repoNames)
```
Comment (varungandhi-src) [inline internal/gitserver/client.go:786]:
```suggestion
var repoNames []string
for _, rc := range repoCommits {
	if _, ok := repoNameMap[rc.Repo]; !ok {
		repoNameMap[rc.Repo] = struct{}{}
		repoNames = append(repoNames, string(rc.Repo))  
    }
}
```
1. Determinism without sorting.
2. IIUC the common case ought to be that most of the commits are for the same repo, so using `var repoNames []string` instead of `make` (but you could change that if I'm wrong).
3. Fusing the loops avoids iterating twice.

File: internal/gitserver/client.go [hunk 9]
```diff
@@ -614,6 +631,163 @@ var deadlineExceededCounter = promauto.NewCounter(prometheus.CounterOpts{
...
+	defer endObservation(1, observation.Args{})
+
+	// Make a request to a signle gitserver shard and feed the results to the user-supplied
+	// callback. Invoked multiple times (and concurrently) in the function defined in the
+	// following.
+	performLogRequestToShard := func(addr string, repoCommits []api.RepoCommit) (err error) {
+		var numProcessed int
+		repoNames := repoNamesFromRepoCommits(repoCommits)
+
+		ctx, logger, endObservation := c.operations.batchLogSingle.WithAndLogger(ctx, &err, observation.Args{
```
Comment (varungandhi-src) [inline internal/gitserver/client.go:651]: Q: Is capturing the `Context` value by reference the idiomatic thing or should it be passed in as a parameter?

File: internal/gitserver/client.go [hunk 10]
```diff
@@ -614,6 +631,163 @@ var deadlineExceededCounter = promauto.NewCounter(prometheus.CounterOpts{
...
+		resp, err := c.do(ctx, repoName, "POST", uri, buf.Bytes())
+		if err != nil {
+			return err
+		}
+		defer resp.Body.Close()
+		logger.Log(log.Int("resp.StatusCode", resp.StatusCode))
+
+		if resp.StatusCode != http.StatusOK {
+			body, _ := io.ReadAll(io.LimitReader(resp.Body, 200))
+			return errors.Newf("http status %d: %s", resp.StatusCode, body)
```
Comment (varungandhi-src) [inline internal/gitserver/client.go:688]: Q: What's the reason behind using `io.LimitReader` here? Is the concern that there will be a bunch of duplication in the error message, with a bunch of it being useless?

Comment (efritz) [reply internal/gitserver/client.go:688]: Could be unbounded in size, but we want to know at least what the initial error message might be.

File: internal/gitserver/client.go [hunk 11]
```diff
@@ -614,6 +631,163 @@ var deadlineExceededCounter = promauto.NewCounter(prometheus.CounterOpts{
...
+		var response protocol.BatchLogResponse
+		if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
+			return err
+		}
+		logger.Log(log.Int("numResults", len(response.Results)))
+
+		for _, result := range response.Results {
+			var err error
+			if result.CommandError != "" {
+				err = errors.Newf(result.CommandError)
```
Comment (varungandhi-src) [inline internal/gitserver/client.go:700]:
```suggestion
				err = errors.New(result.CommandError)
```

To avoid (mis-)interpreting any formatting specifiers accidentally present in `result.CommandError`.

File: internal/gitserver/client.go [hunk 12]
```diff
@@ -614,6 +631,163 @@ var deadlineExceededCounter = promauto.NewCounter(prometheus.CounterOpts{
...
+
+	batches := make(map[string][]api.RepoCommit, len(addrsByName))
+	for _, repoCommit := range opts.RepoCommits {
+		addr := addrsByName[repoCommit.Repo]
+
+		batches[addr] = append(batches[addr], api.RepoCommit{
+			Repo:     repoCommit.Repo,
+			CommitID: repoCommit.CommitID,
+		})
+	}
```
Comment (varungandhi-src) [inline internal/gitserver/client.go:748]: suggestion: Perhaps these loops should be fused too? Something like:

```
	addrsByName := make(map[api.RepoName]string, len(opts.RepoCommits))
	var batches map[string][]api.RepoCommit 
	for _, repoCommit := range opts.RepoCommits {
		var addr string
        var ok bool 
		if addr, ok = addrsByName[repoCommit.Repo]; !ok {
			if addr, err = c.AddrForRepo(ctx, repoCommit.Repo); err != nil {
				return err
			}
			addrsByName[repoCommit.Repo] = addr
		}
		batches[addr] = append(batches[addr], api.RepoCommit{
			Repo:     repoCommit.Repo,
			CommitID: repoCommit.CommitID,
		})
	}
```

Right now the second loop isn't self-contained because making sure that the lookup below will always succeed:
```
addr := addrsByName[repoCommit.Repo]
```
requires looking at the body of the first look.

File: internal/gitserver/client.go (hunk 9 above)
Comment (efritz) [reply internal/gitserver/client.go:651]: 1. No it's probably not
2. I was using the wrong context before this.

Good catch! Will update.

File: cmd/gitserver/server/server.go (hunk 2 above)
Comment (efritz) [reply cmd/gitserver/server/server.go:1154]: Will update.

File: enterprise/internal/codeintel/gitserver/client.go (hunk 7 above)
Comment (efritz) [reply enterprise/internal/codeintel/gitserver/client.go]: This is only true on an earlier commit - check where we actually call out to the vcs package.

File: internal/gitserver/client.go (hunk 8 above)
Comment (efritz) [reply internal/gitserver/client.go:786]: 1/3. Can do that.
2. Common case might be all *different* repos, if we're talking about find ref operations.

Comment (efritz) [reply internal/gitserver/client.go:786]: Updated, PTAL.

File: internal/gitserver/client.go (hunk 12 above)
Comment (efritz) [reply internal/gitserver/client.go:751]: Updated, doesn't actually seem bad.

File: internal/api/api.go [hunk 13]
```diff
@@ -39,6 +41,19 @@ func (c CommitID) Short() string {
...
+// RepoCommit scopes a commit to a repository.
+type RepoCommit struct {
+	Repo     RepoName
+	CommitID CommitID
+}
+
+func (rc RepoCommit) LogFields() []log.Field {
+	return []log.Field{
+		log.String("repo", string(rc.Repo)),
+		log.String("rommitID", string(rc.CommitID)),
```
Comment (efritz) [inline internal/api/api.go:53]:
```suggestion
		log.String("commitID", string(rc.CommitID)),
```

File: enterprise/internal/codeintel/gitserver/client.go (hunk 7 above)
Comment (varungandhi-src) [reply enterprise/internal/codeintel/gitserver/client.go]: Gotcha', yeah that makes sense. I hadn't looked at the later commits when I left this comment.

File: internal/vcs/git/commits.go [hunk 14]
```diff
@@ -526,6 +532,103 @@ func CommitExists(ctx context.Context, db database.DB, repo api.RepoName, id api
...
+				return nil
+			}
+			return errors.Wrap(err, "parseCommitLogOutput")
+		}
+
+		// Enforce sub-repository permissions
+		filteredCommits, err := filterCommits(ctx, wrappedCommits, repoCommit.Repo, checker)
+		if err != nil {
+			return errors.Wrap(err, "filterCommits")
+		}
```
Comment (varungandhi-src) [inline internal/vcs/git/commits.go:603]: There's no check for `ignoreErrors` here. If that's deliberate, would be useful to add a short comment for why that's the case.

File: internal/vcs/git/commits.go [hunk 15]
```diff
@@ -526,6 +532,103 @@ func CommitExists(ctx context.Context, db database.DB, repo api.RepoName, id api
...
+		if len(filteredCommits) == 0 {
+			// Not found
+			return nil
+		}
```
Comment (varungandhi-src) [inline internal/vcs/git/commits.go:607]: Q: Is it an error if `len(filteredCommits) > 1` / should that be logged/bubbled up?

File: internal/vcs/git/commits.go [hunk 16]
```diff
@@ -521,6 +527,103 @@ func CommitExists(ctx context.Context, db database.DB, repo api.RepoName, id api
...
+				// Treat as not-found
+				return nil
+			}
+			return errors.Wrap(err, "parseCommitLogOutput")
+		}
+
+		// Enforce sub-repository permissions
+		filteredCommits, err := filterCommits(ctx, wrappedCommits, repoCommit.Repo, checker)
+		if err != nil {
+			return errors.Wrap(err, "filterCommits")
```
Comment (efritz) [inline internal/vcs/git/commits.go:602]:
```suggestion
			// Note that we don't check ignoreErrors on this condition. When we
			// ignore errors it's to hide an issue with a single git log request on a
			// single shard, which could return an error if that repo is missing, the
			// supplied commit does not exist in the clone, or if the repo is malformed.
			//
			// We don't want to hide unrelated infrastructure errors caused by this
			// method call.
			return errors.Wrap(err, "filterCommits")
```

File: internal/vcs/git/commits.go (hunk 14 above)
Comment (efritz) [reply internal/vcs/git/commits.go:603]: https://github.com/sourcegraph/sourcegraph/pull/33195/files#r839953322

File: internal/vcs/git/commits.go (hunk 15 above)
Comment (efritz) [reply internal/vcs/git/commits.go:614]: It looks like that's what `gitCommit` does. It's probably good to add a check for it. I'll make sure we add that condition explicitly.

File: enterprise/internal/codeintel/gitserver/client.go [hunk 17]
```diff
@@ -75,19 +75,17 @@ func (c *Client) CommitsExist(ctx context.Context, commits []RepositoryCommit) (
...
-		}
-
-		e, err := git.CommitExists(ctx, c.db, name, api.CommitID(rc.Commit), authz.DefaultSubRepoPermsChecker)
-		if err != nil {
-			return nil, err
-		}
+	repoCommits := make([]api.RepoCommit, 0, len(commits))
+	for _, rc := range commits {
+		repoCommits = append(repoCommits, api.RepoCommit{
+			Repo:     repositoryNames[rc.RepositoryID],
```
Comment (varungandhi-src) [inline enterprise/internal/codeintel/gitserver/client.go:81]: Q: Do you think it is worth guarding against an invariant failure here where the `DBStore`'s `RepoNames` method is incorrectly implemented, returning a map which doesn't cover all the repository IDs? If so, perhaps we should return an error here if the map access fails.

**Review by varungandhi-src — APPROVED** [review]
LGTM; the implementation is clean and understandable. Not 100% on the testing front (e.g. if the existing tests provide good coverage or if there are edge cases that we should cover with more tests), so no comments on that.

File: enterprise/internal/codeintel/gitserver/client.go (hunk 17 above)
Comment (varungandhi-src) [reply enterprise/internal/codeintel/gitserver/client.go:81]: Non-blocking suggestion: For this point specifically, I was thinking of something like `if repo, ok := repositoryNames[rc.RepositoryID]; !ok { /* return error */ }`, but yeah, handling it at the lower layer (probably at the right place) would require some more changes.

Comment (efritz) [reply enterprise/internal/codeintel/gitserver/client.go:81]: Oh that's probably a decent solution if we need to get this in before doing it in a way that keeps partial results in mind. I think either way has an error will blow up this particular method call, so we won't lose anything with an interim solution.

Whatever I end up doing for this PR I'll make sure we have a follow-up issue if necessary.

File: internal/vcs/git/commits.go (hunk 15 above)
Comment (efritz) [reply internal/vcs/git/commits.go:614]: Updated.

**Comment by efritz** [comment]
> Not 100% sure on the testing front

I'm fairly confident on decent coverage from new unit tests for some of the basic operation and edge cases. Code Intel QA test does exercise this entire path as well so we know it's hooked up correctly.

File: enterprise/internal/codeintel/gitserver/client.go (hunk 17 above)
Comment (efritz) [reply enterprise/internal/codeintel/gitserver/client.go]: @varungandhi-src Can you take a look at ae8e6e506f89e374c8b570e86efa3dc0fc6ec0fb?

File: enterprise/internal/codeintel/gitserver/client.go [hunk 18]
```diff
@@ -87,16 +95,47 @@ func (c *Client) CommitsExist(ctx context.Context, commits []RepositoryCommit) (
...
+	return pushFalseToSliceAtIndexes(exists, indexesWithMissingRepos)
+}
+
+// pushFalseToSliceAtIndexes returns a copy of the input slice with the value false inserted into each
+// of the indexes. This has the effect of sliding all of the subsequent elements down one index on every
+// insertion.
+//
+// This function assumes that indexes are within the range [0, len(exists)+len(indexes)) and are given
+// in ascending order. See CommitsExists for its exact usage.
+func pushFalseToSliceAtIndexes(exists []bool, indexes []int) ([]bool, error) {
```
Comment (varungandhi-src) [inline enterprise/internal/codeintel/gitserver/client.go:117]: Suggestion: If I'm understanding the code correctly, I think it would be much simpler to keep track of indices which `exists` corresponds to, instead of the current logic, avoiding the need to push/slide anything.

1. When appending to `repoCommits`, add the index to a slice: `indexesForCommit`. So `len(indexesForCommit) == len(repoCommits)`.
2. When we get `exists`, iterate over `repoCommitIndexes` and store the `exists` values at the right indexes.
```
existsFinal := make([]bool, len(commits)) // zero/false-initialize deliberately
// pre-condition len(repoCommits) == len(exists) == len(indexesForCommit)
for i, indexForCommit := range indexesForCommit {
    existsFinal[indexForCommit] = exists[i]
}
return existsFinal, nil
```
(Feel free to fix the naming in the code example)

Comment (efritz) [reply enterprise/internal/codeintel/gitserver/client.go:117]: That's way simpler 🤦

**Comment by ryanslade** [comment]
Thanks for this, I haven't forgotten to take a look but our whole team are occupied by a customer issue at the moment.

**Review by andreeleuterio — APPROVED** [review]
LGTM. The only suggestion is to return an error on non-POST requests, as it helps to harden against SSRF. This can be done as a later improvement and is not blocking to ship the feature.

**Comment by efritz** [comment]
@andreeleuterio Done.

File: cmd/gitserver/server/server.go [hunk 19]
```diff
@@ -1133,6 +1142,124 @@ func matchCount(cm *protocol.CommitMatch) int {
...
+		var req protocol.BatchLogRequest
+		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+			return http.StatusBadRequest, err
+		}
+		logger.Log(req.LogFields()...)
+
+		// Validate request parameters
+		if len(req.RepoCommits) == 0 {
+			// Early exit
+			w.WriteHeader(http.StatusOK)
```
Comment (ryanslade) [inline cmd/gitserver/server/server.go:1207]: nit: This is automatically written once you start sending the response.

File: cmd/gitserver/server/server.go [hunk 20]
```diff
@@ -1133,6 +1142,124 @@ func matchCount(cm *protocol.CommitMatch) int {
...
+
+			results = append(results, protocol.BatchLogResult{
+				RepoCommit:    repoCommit,
+				CommandOutput: output,
+				CommandError:  errMessage,
+			})
+		}
+
+		// Write payload to client
+		w.WriteHeader(http.StatusOK)
```
Comment (ryanslade) [inline cmd/gitserver/server/server.go:1239]: nit: This is automatically written once you start sending the response.

File: internal/gitserver/client.go [hunk 21]
```diff
@@ -608,6 +637,153 @@ var deadlineExceededCounter = promauto.NewCounter(prometheus.CounterOpts{
...
+// BatchLog invokes the given callback with the `git log` output for a batch of repository
+// and commit pairs. If the invoked callback returns a non-nil error, the operation will begin
+// to abort processing further results.
+func (c *ClientImplementor) BatchLog(ctx context.Context, opts BatchLogOptions, callback BatchLogCallback) (err error) {
+	ctx, endObservation := c.operations.batchLog.With(ctx, &err, observation.Args{LogFields: opts.LogFields()})
+	defer endObservation(1, observation.Args{})
+
+	// Make a request to a signle gitserver shard and feed the results to the user-supplied
+	// callback. Invoked multiple times (and concurrently) in the function defined in the
+	// following.
```
Comment (ryanslade) [inline internal/gitserver/client.go:649]: Looks like you missed part of the sentence?

File: internal/gitserver/client.go [hunk 22]
```diff
@@ -608,6 +637,153 @@ var deadlineExceededCounter = promauto.NewCounter(prometheus.CounterOpts{
...
+		})
+	}
+
+	// Perform each batch reqeuest concurrently, but only allow four requests to be in-flight at a
+	// time. This operation returns partial results in the case of a malformed or missing repository
+	// or a bad commit reference, but does not attempt to return partial results when an entire shard
+	// is down. Any of these operations failing will cause an error to be returned from the entire
+	// BatchLog function.
+
+	sem := semaphore.NewWeighted(int64(4))
```
Comment (ryanslade) [inline internal/gitserver/client.go:750]:
```suggestion
	sem := semaphore.NewWeighted(4)
```

File: internal/vcs/git/commits_test.go [hunk 23]
```diff
@@ -1189,6 +1189,83 @@ func TestCommitDate(t *testing.T) {
...
+			nil,
+		}
+		if diff := cmp.Diff(expectedCommits, commits); diff != "" {
+			t.Errorf("unexpected commits (-want +got):\n%s", diff)
+		}
+	})
+	t.Run("with sub-repo permissions", func(t *testing.T) {
+		t.Skip()
+
+		// Note: I'm not sure if the sub-repo permissions checker here does anything, or if I'm just
```
Comment (ryanslade) [inline internal/vcs/git/commits_test.go:1263]: Correct, sub-repo permissions on commits are applied based on whether all files in the commit are visible to the current user. 

To test you'd need to return some files and use a mock checker which can be created using `getTestSubRepoPermsChecker`

**Review by ryanslade — APPROVED** [review]
Nice!

**Comment by ryanslade** [comment]
And thanks for improving our observability in the process :)

File: cmd/gitserver/server/server.go [hunk 24]
```diff
@@ -1133,6 +1142,124 @@ func matchCount(cm *protocol.CommitMatch) int {
 	return 1
 }
 
+func (s *Server) handleBatchLog(w http.ResponseWriter, r *http.Request) {
+	// 🚨 SECURITY: Only allow POST requests.
+	if strings.ToUpper(r.Method) != "POST" {
```
[the remark this answers is not in the document]
Comment (efritz) [reply cmd/gitserver/server/server.go:1147]: Copied from another part in this file. Updated both places.

File: cmd/gitserver/server/server.go (hunk 20 above)
Comment (efritz) [reply cmd/gitserver/server/server.go:1239]: Updated comment, removed line.

File: internal/gitserver/client.go (hunk 21 above)
Comment (efritz) [reply internal/gitserver/client.go:649]: Rewrote to make more clear - also fixed typo.
